README.md
# NN
NTechLab assignment

## Description
* Rest service to find kNN in R radius among 1M records.
* Service supports CRUD operations with records
 
## Requirements
* Python language
* virtualenv for reproduction

## Implementation details
* Flask framework to process http requests
* SQLAlchemy and sqlite for storing data
* Unit and inegration tests

## Server deployment/cleanup

### Install project
* pip install virtualenv
* git clone https://github.com/DrNecromant/NN
* virtualenv NN
* cd NN
* . bin/activate
* pip install -r requirements.txt

### Start application
* python main.py

### Stop application
* CTRL+C

### Exit and remove project
* deactivate
* cd ..
* rm -r NN

## Usage and examples

### API usage

#### Create users
```
$curl http://127.0.0.1:5000/v1/NN/users -X POST -d '{"x": 1, "y": 2}'
{
    "message": "Created", 
    "user_url": "http://127.0.0.1:5000/v1/NN/users/1"
}
$curl http://127.0.0.1:5000/v1/NN/users -X POST -d '{"x": 2, "y": 1}'
{
    "message": "Created", 
    "user_url": "http://127.0.0.1:5000/v1/NN/users/2"
}
$curl http://127.0.0.1:5000/v1/NN/users -X POST -d '{"x": 0, "y": 0}'
{
    "message": "Created", 
    "user_url": "http://127.0.0.1:5000/v1/NN/users/3"
}
```

#### Modify and delete users
```
$curl http://127.0.0.1:5000/v1/NN/users/1 -X POST -d '{"y": 3}'
{
    "message": "OK", 
    "user_url": "http://127.0.0.1:5000/v1/NN/users/1/1"
}
$curl http://127.0.0.1:5000/v1/NN/users/3 -X DELETE
{
    "message": "OK"
}
```

#### Show users and stats info
```
$curl http://127.0.0.1:5000/v1/NN/users -X GET
{
    "message": "OK", 
    "users": {
        "1": {
            "user_url": "http://127.0.0.1:5000/v1/NN/users/1", 
            "x": 1, 
            "y": 3
        }, 
        "2": {
            "user_url": "http://127.0.0.1:5000/v1/NN/users/2", 
            "x": 2, 
            "y": 1
        }
    }
}
$curl http://127.0.0.1:5000/v1/NN/users/info -X GET
{
    "message": "OK", 
    "user_count": 2
}
```

#### find kNN
```
$curl http://127.0.0.1:5000/v1/NN/users/knn?U=1\&R=5 -X GET
{
    "message": "OK", 
    "result": 1
}
$curl http://127.0.0.1:5000/v1/NN/users/knn?U=1\&R=1 -X GET
{
    "message": "OK", 
    "result": 0
}
```

### Unittests
```
$python unittests.py
testDBUser (__main__.TestDB) ... ok
testDBUserStats (__main__.TestDB) ... ok
testAddInvalidUser (__main__.TestUserList) ... ok
testAddUser (__main__.TestUserList) ... ok
testAddiExistedUser (__main__.TestUserList) ... ok
testEmptyUserList (__main__.TestUserList) ... ok
testFullUserList (__main__.TestUserList) ... ok
testUserListPages (__main__.TestUserList) ... ok
testDeleteUser (__main__.TestUser) ... ok
testGetUser (__main__.TestUser) ... ok
testUpdateUser (__main__.TestUser) ... ok
testGetInfo (__main__.TestInfo) ... ok
testGetKnn (__main__.TestKnn) ... ok

----------------------------------------------------------------------
Ran 13 tests in 0.475s
OK
```

### Integration tests
```
$python integrationtests.py 
Generage 10000 users...
Get kNN by binary algorythm...
Get kNN by calculating all distances...
Results are equal 2559 2559

----------------------------------------------------------------------
Ran 1 test in 185.561s
OK
```

### Benchmark tests
```
$python benchmark.py 

test | attempt_id | radius | knn | exec_time

testBinarySearch 1 10 14 0.586269855499 
testBinarySearch 2 10 18 0.828803062439 
testBinarySearch 3 10 15 0.549397945404 
testBinarySearch 1 50 322 3.55142521858 
testBinarySearch 2 50 309 4.53496789932 
testBinarySearch 3 50 320 3.75467896461 
testBinarySearch 1 100 1264 5.76161003113 
testBinarySearch 2 100 1267 6.04228281975 
testBinarySearch 3 100 1243 5.056016922 
testBinarySearch 1 500 29780 4.59962797165 
testBinarySearch 2 500 31216 5.4956099987 
testBinarySearch 3 500 25684 3.68939399719 
testBinarySearch 1 1000 93543 5.79501605034 
testBinarySearch 2 1000 119628 6.90547990799 
testBinarySearch 3 1000 80905 3.77246212959 
testAllDistanceSearch 1 10 14 14.5994999409 
testAllDistanceSearch 2 10 18 14.5038819313 
testAllDistanceSearch 3 10 15 14.7285120487 
testAllDistanceSearch 1 50 322 14.439538002 
testAllDistanceSearch 2 50 309 14.9254710674 
testAllDistanceSearch 3 50 320 14.4932551384 
testAllDistanceSearch 1 100 1264 14.7746219635 
testAllDistanceSearch 2 100 1267 14.5922350883 
testAllDistanceSearch 3 100 1243 14.7975559235 
testAllDistanceSearch 1 500 29780 14.7092351913 
testAllDistanceSearch 2 500 31216 15.002371788 
testAllDistanceSearch 3 500 25684 14.5927751064 
testAllDistanceSearch 1 1000 93543 15.8682169914 
testAllDistanceSearch 2 1000 119628 15.1586270332 
testAllDistanceSearch 3 1000 80905 14.8813381195 

RESULTS:
	testAllDistanceSearch: 14.8044756889
	testBinarySearch: 4.06153618495
```

## TODO
* Info controller should provide also max, min, avg, density stats
* All math operations should be moved to separate module
* Unittests for math operations
* Add good logging, remove/replace prints to support Python 3 (!!!)
* Algorythm optimization: choose carefully size of rectangle to split

## License
Feel free to use my code on production environments and good luck. =)


CHANGELOG.md
# Change Log

## v1.3.1
- Added documenation for usage
- Added TODO
- Fixed DELETE operation
- Minor fixes for comments and logs
- Cleanup

## v1.3.0
- Added benchmark test
- Added pre-created benchmark DB
- Integration test optimization
- Main algorythm optimization

## v1.2.3
- Added integration test

## v1.2.2
- Re-create DB on service start
- Add comments to Info controller
- Add requests module to requirements

## v1.2.1
- Minor fixes and improvements
- Cleanup

## v1.2.0
- Main algorythm implementation
- Minor fixes

## v1.1.5
- Initial steps in main algorythm, preparations

## v1.1.4
- Exclude sum from DBUserStats

## v1.1.3
- Fix unittest for DBUserStats

## v1.1.2
- Extend unittests for DBUserStats
- Fix DBUserStats with limit and offsets

## v1.1.1
- Add user stats model
- Unitests for userstats
- Minor fixes and comments

## v1.1.0
- kNN controller skeleton
- unittests for kNN controller

## v1.0.5
- Add User resource with unittest
- Fixed testInfo

## v1.0.4
- Minor fix: uri for test sql base

## v1.0.3
- Add pages functionality to list users

## v1.0.2
- Add resource info that returns users count

## v1.0.1
- UserList unitets improvement

## v1.0.0
- UserList with unittests implementation

## v0.2.0
- Skeleton for main app (CRUD)
- Added conts module with app config
- Added Sql module
- Updated requirements

## v0.1.1
- Added missed requiroments

## v0.1.0
- Initial commit
- Initial documentation
- Add pip requirements


models.py
from flask_sqlalchemy import SQLAlchemy

db = SQLAlchemy()

class DBUser(db.Model):
	"""
	Users with their coordinates
	"""
	id = db.Column(db.Integer, primary_key = True)
	x = db.Column(db.Integer)
	y = db.Column(db.Integer)

	def __init__(self, x, y):
		self.x = x
		self.y = y

class DBUserStats(object):
	"""
	Get stats from DBUser such as:
	min, max, average, total with where clause
	"""

	def __init__(self, offsetX = None, offsetY = None, \
				limitX = None, limitY = None):
		query = db.session.query(
			db.func.min(DBUser.x).label("minX"), \
			db.func.min(DBUser.y).label("minY"), \
			db.func.max(DBUser.x).label("maxX"), \
			db.func.max(DBUser.y).label("maxY"), \
			db.func.avg(DBUser.x).label("avgX"), \
			db.func.avg(DBUser.y).label("avgY"), \
			db.func.count(DBUser.id).label("count"), \
		)

		if offsetX is not None:
			query = query.filter(DBUser.x >= offsetX)
		if offsetY is not None:
			query = query.filter(DBUser.y >= offsetY)
		if limitX is not None:
			query = query.filter(DBUser.x <= limitX)
		if limitY is not None:
			query = query.filter(DBUser.y <= limitY)

		self.result = query.one()

	@property
	def rect(self):
		"""
		return rect corner coordinates
		"""
		return (
			(self.result.minX, self.result.minY), \
			(self.result.minX, self.result.maxY), \
			(self.result.maxX, self.result.minY), \
			(self.result.maxX, self.result.maxY), \
		)

	@property
	def minX(self):
		return self.result.minX

	@property
	def minY(self):
		return self.result.minY

	@property
	def maxX(self):
		return self.result.maxX

	@property
	def maxY(self):
		return self.result.maxY

	@property
	def avgX(self):
		return int(self.result.avgX)

	@property
	def avgY(self):
		return int(self.result.avgY)

	@property
	def count(self):
		return self.result.count


benchmark.py
import unittest
import random
from math import sqrt
import time
from pprint import pprint

from main import app
from flask_api import status

from consts import *
from models import *

db.app = app
db.init_app(app)
client = app.test_client()

# define initial variables
initial_user_id_list = [1, 2, 3]
radius_list = [10, 50, 100 , 500, 1000]
baseurl = "v1/NN"

def testAllDistanceSearch(radius, initial_user_id):
	url = "%s/users/knn?R=%s&U=%s&dist=Y" % (baseurl, radius, initial_user_id)
	res = client.get(url)
	return eval(res.get_data())["result"]

def testBinarySearch(radius, initial_user_id):
	url = "%s/users/knn?R=%s&U=%s" % (baseurl, radius, initial_user_id)
	res = client.get(url)
	return eval(res.get_data())["result"]

if __name__ == '__main__':
	results = dict()
	print "\ntest | attempt_id | radius | knn | exec_time"
	for test in (testBinarySearch, testAllDistanceSearch):
		timelist = list()
		for radius in radius_list:
			for initial_user_id in initial_user_id_list:
				init_time = time.time()
				knn = test(radius, initial_user_id)
				exec_time = time.time() - init_time
				timelist.append(exec_time)
				print "\n", test.__name__, initial_user_id, radius, knn, exec_time, 
		results[test.__name__] = sum(timelist) / len(timelist)
	print "\n\nRESULTS:\n"
	for testname in results:
		print "\t%s: %s" %(testname, results[testname])


unittests.py
import unittest
import random
from math import sqrt

from main import app
from flask_api import status

from consts import *
from models import *

db.app = app
db.init_app(app)
db.create_all()

# generate random data
coord = [(x, y) for x in xrange(1, 1000) for y in xrange(1, 1000)]
random.shuffle(coord)

class TestDB(unittest.TestCase):
	"""
	Unittests for DB
	"""

	def _assertValidStats(self, stats, testdata):
		"""
		Compare DB stats and list stats
		"""
		xlist, ylist = zip(*testdata)
		self.assertEquals(stats.minX, min(xlist))
		self.assertEquals(stats.minY, min(ylist))
		self.assertEquals(stats.maxX, max(xlist))
		self.assertEquals(stats.maxY, max(ylist))
		self.assertEquals(stats.count, len(testdata))

	def testDBUser(self):
		"""
		Try to create record in a table DBUser
		"""
		user = DBUser(*coord.pop())
		db.session.add(user)
		db.session.commit()

	def testDBUserStats(self):
		"""
		Generate test data and validate DBUserStats
		"""
		# Generate testdata from random sequences
		testdata = [coord.pop() for i in range(SQL_TESTDATA_COUNT)]

		# Fill table User with testdata
		DBUser.query.delete()
		for x, y in testdata:
			user = DBUser(x, y)
			db.session.add(user)
		db.session.commit()

		# Get stats from DB and validate them
		stats = DBUserStats()
		self._assertValidStats(stats, testdata)

		# Test stats with limits
		# Cut testdata for that
		sorted_x, sorted_y = map(sorted, zip(*testdata))
		cut_index = int(sqrt(SQL_TESTDATA_COUNT) / 5)
		offsetX = sorted_x[cut_index]
		offsetY = sorted_y[cut_index]
		limitX = sorted_x[-cut_index]
		limitY = sorted_y[-cut_index]
		testdata = [
			(x, y) for x, y in testdata \
			if y <= limitY and y >= offsetY and \
			x <= limitX and x >= offsetX
		]

		# Get stats from DB and validate them
		stats = DBUserStats(offsetX, offsetY, limitX, limitY)
		self._assertValidStats(stats, testdata)

class TestUserList(unittest.TestCase):
	"""
	Unittests for UserList
	"""
	def setUp(self):
		self.client = app.test_client()
		self.url = "%s/users" % BASEURL

	def testAddUser(self):
		"""
		Try to add new user
		Check return code is 201
		"""
		res = self.client.post(self.url, data = '{"x": %s, "y": %s}' % coord.pop())
		self.assertEquals(res.status_code, status.HTTP_201_CREATED)

	def testAddInvalidUser(self):
		"""
		Try to add new user without x or y
		Check return code is 400
		"""
		c = coord.pop()
		res = self.client.post(self.url, data = '{"x": %s}' % c[0])
		self.assertEquals(res.status_code, status.HTTP_400_BAD_REQUEST)
		res = self.client.post(self.url, data = '{"y": %s}' % c[1])
		self.assertEquals(res.status_code, status.HTTP_400_BAD_REQUEST)

	def testAddiExistedUser(self):
		"""
		Try to add new existed user
		Check return code is 409
		"""
		c = coord.pop()
		self.client.post(self.url, data = '{"x": %s, "y": %s}' % c)
		res = self.client.post(self.url, data = '{"x": %s, "y": %s}' % c)
		self.assertEquals(res.status_code, status.HTTP_409_CONFLICT)

	def testEmptyUserList(self):
		"""
		Try to get empty UserList
		Check return code is 400
		"""
		DBUser.query.delete()
		res = self.client.get(self.url)
		self.assertEquals(res.status_code, status.HTTP_404_NOT_FOUND)

	def testFullUserList(self):
		"""
		Try to get UserList
		Check return code is 200
		"""
		self.client.post(self.url, data = '{"x": %s, "y": %s}' % coord.pop())
		res = self.client.get(self.url)
		self.assertEquals(res.status_code, status.HTTP_200_OK)

	def testUserListPages(self):
		"""
		Remove all data, generate 10 records.
		Try to show data for page=1 with pagesize=9
		Result code 200 is expected
		Then try to move over maximum
		with page=0, pagesize=10 and page=2, pagesize=9
		Result code 404 is expected
		"""
		DBUser.query.delete()
		for i in range(10):
			self.client.post(self.url, data = '{"x": %s, "y": %s}' % coord.pop())

		url_params = "?page=1&pagesize=9"
		res = self.client.get(self.url + url_params)
		self.assertEquals(res.status_code, status.HTTP_200_OK)

		url_params = "?page=2&pagesize=9"
		res = self.client.get(self.url + url_params)
		self.assertEquals(res.status_code, status.HTTP_404_NOT_FOUND)

		url_params = "?page=1&pagesize=10"
		res = self.client.get(self.url + url_params)
		self.assertEquals(res.status_code, status.HTTP_404_NOT_FOUND)

class TestUser(unittest.TestCase):
	"""
	Unittests for User: get, update, delete
	"""

	def setUp(self):
		self.client = app.test_client()
		self.url = "%s/users" % BASEURL

	def _get_user_url(self, user_id):
		return "%s/%s" % (self.url, user_id)

	def _create_db_user(self):
		"""
		Create DB user and return user object
		"""
		user = DBUser(*coord.pop())
		db.session.add(user)
		db.session.flush()
		db.session.commit()
		return user

	def testGetUser(self):
		"""
		Add user into DB, try to get it by id: 200 is expected
		After that try to get user by id + 100: 404 is expected
		"""
		user = self._create_db_user()
		res = self.client.get(self._get_user_url(user.id))
		self.assertEquals(res.status_code, status.HTTP_200_OK)
		res = self.client.get(self._get_user_url(user.id + 100))
		self.assertEquals(res.status_code, status.HTTP_404_NOT_FOUND)

	def testUpdateUser(self):
		"""
		Check parameters might be updated one by one
		Compare updated parameters with old ones
		Validate that request without parameters is invalid
		"""
		user = self._create_db_user()
		user_id = user.id
		url = self._get_user_url(user_id)
		old_x, old_y = user.x, user.y

		c = coord.pop()
		res = self.client.post(url, data = '{"x": %s}' % c[0])
		self.assertEquals(res.status_code, status.HTTP_200_OK)
		res = self.client.post(url, data = '{"y": %s}' % c[1])
		self.assertEquals(res.status_code, status.HTTP_200_OK)
		res = self.client.post(url, data = '{}')
		self.assertEquals(res.status_code, status.HTTP_400_BAD_REQUEST)

		user = DBUser.query.filter_by(id = user_id).one()
		self.assertNotEqual((old_x, old_y), (user.x, user.y))

	def testDeleteUser(self):
		"""
		Add new user in DB and delete it with API
		Check there is no such user in DB anymore
		"""
		user = self._create_db_user()
		user_id = user.id
		url = self._get_user_url(user_id)
		res = self.client.delete(url)
		self.assertEquals(res.status_code, status.HTTP_200_OK)

		user = DBUser.query.filter_by(id = user_id).first()
		self.assertIsNone(user)

class TestInfo(unittest.TestCase):
	"""
	Unittests for Info
	"""
	def setUp(self):
		self.client = app.test_client()
		self.url = "%s/users/info" % BASEURL

	def testGetInfo(self):
		"""
		Try to get info
		Check return code is 200
		"""
		res = self.client.get(self.url)
		self.assertEquals(res.status_code, status.HTTP_200_OK)

class TestKnn(unittest.TestCase):
	"""
	Unittests for Knn
	"""
	def setUp(self):
		self.radius = SQL_TESTDATA_COUNT / 3
		self.user_id = 1
		self.client = app.test_client()
		self.url = "%s/users/knn" % BASEURL

	def testGetKnn(self):
		"""
		Check request without mandatory arguments
		Check Knn return 200 with mandatory arguments
		"""
		Rarg = "R=%s" % self.radius
		Uarg = "U=%s" % self.user_id
		for arg in (Rarg, Uarg):
			res = self.client.get("%s?%s" % (self.url, arg))
			self.assertEquals(res.status_code, status.HTTP_400_BAD_REQUEST)

		DBUser.query.delete()
		for i in range(SQL_TESTDATA_COUNT):
			db.session.add(DBUser(*coord.pop()))
		db.session.commit()
		res = self.client.get("%s?%s&%s" % (self.url, Rarg, Uarg))
		self.assertEquals(res.status_code, status.HTTP_200_OK)
		res = self.client.get("%s?%s&%s&dist=Y" % (self.url, Rarg, Uarg))
		self.assertEquals(res.status_code, status.HTTP_200_OK)

if __name__ == "__main__":
	suites = list()
	for test in (TestDB, TestUserList, TestUser, TestInfo, TestKnn):
		suites.append(unittest.TestLoader().loadTestsFromTestCase(test))
	suite = unittest.TestSuite(suites)
	results = unittest.TextTestRunner(verbosity = 2).run(suite)


integrationtests.py
import unittest
import random
from math import sqrt
import requests

from main import app
from flask_api import status

from consts import *

# generate random data
coord = [(x, y) for x in xrange(1, 1000) for y in xrange(1, 1000)]
random.shuffle(coord)

class TestNN(unittest.TestCase):
	"""
	Integration test for NN
	"""
	def setUp(self):
		self.baseurl = "http://127.0.0.1:5000/v1/NN"
		self.user_count = 10000
		self.radius = 300
		self.user_id = 1
		self.page_size = 238

	def testCheckAlgorythm(self):
		"""
		Add 10k data with rest
		Get kNN with binary algorythm
		Calculate all distances manually
		Compare results
		"""
		# generate data
		print "Generage %s users..." % self.user_count
		url = "%s/users" % self.baseurl
		for i in range(self.user_count):
			res = requests.post(url, data = '{"x": "%s", "y": "%s"}' % coord.pop())
			self.assertEqual(res.status_code, status.HTTP_201_CREATED)
			
		Rarg = "R=%s" % self.radius
		Uarg = "U=%s" % self.user_id
		Darg = "dist=Y"

		print "Get kNN by binary algorythm..."
		url = "%s/users/knn" % self.baseurl
		url = "%s?%s&%s" % (url, Rarg, Uarg)
		res = requests.get(url)
		self.assertEquals(res.status_code, status.HTTP_200_OK)
		result1 = res.json()["result"]

		print "Get kNN by calculating all distances..."
		url = "%s/users/knn" % self.baseurl
		url = "%s?%s&%s&%s" % (url, Rarg, Uarg, Darg)
		res = requests.get(url)
		self.assertEquals(res.status_code, status.HTTP_200_OK)
		result2 = res.json()["result"]

		self.assertEquals(result1, result2)
		print "Results are equal", result1, result2
				
if __name__ == '__main__':
    unittest.main()


main.py
from math import sqrt
import os
import sys

from flask import Flask, request
from flask_api import status
from flask_restful import Resource, Api

from consts import *
from models import *

app = Flask("NN")
# Load config for app
if __name__ == '__main__':
	app.config.from_object('consts.ProductionConfig')
	# re-create DB to not conflict with old data
	if os.path.exists(DBFile):
		os.unlink(DBFile)
elif sys.argv[0] == "benchmark.py":
	app.config.from_object('consts.BenchmarkConfig')
else:
	app.config.from_object('consts.TestingConfig')
# Catch all unexpected 404s in json format
api = Api(app, catch_all_404s = True)

# Init DB
db.app = app
db.init_app(app)
db.create_all()

class Info(Resource):
	"""
	Provide information about Users
	Example:
		http://127.0.0.1:5000/v1/NN/users/info -X GET
	"""

	def get(self):
		user_count = DBUser.query.count()
		return {
			"message": "OK",
			"user_count": user_count
		}, status.HTTP_200_OK

class UserList(Resource):
	"""
	Controller to show and extend userlist
	If user exists return 409 Conflict 
	Example:
		curl http://127.0.0.1:5000/v1/NN/users -X POST -d '{"x": 1, "y": 2}'
		curl http://127.0.0.1:5000/v1/NN/users?page=2&pagesize=5 -X GET
	"""

	def get(self):
		"""
		Get user list
		By default show only first 100 records
		page and pagesize are configurable by request args
		"""
		page = int(request.args.get('page', 0))
		pagesize = int(request.args.get('pagesize', 100))
		query = DBUser.query
		query = query.offset(page * pagesize)
		query = query.limit(pagesize)
		users = query.all()
		if not users:
			return {
				"message": "Users not found"
			}, status.HTTP_404_NOT_FOUND

		# Create json for each user in list
		json_users = dict()
		for user in users:
			json_users[user.id] = {
				"x": user.x,
				"y": user.y,
				"user_url": "%s/%s" %(request.url, user.id)
			}

		return {
			"message": "OK",
			"users": json_users
		}, status.HTTP_200_OK

	def post(self):
		"""
		Fetch json data from body ignoring Content-Type header by force flag
		If body is not json flask restfull api automatically handles that
		Store new user in DB and return user url
		"""
		json_data = request.get_json(force = True)
		if "x" not in json_data or "y" not in json_data:
			return {
				"message": "Bad request. x and y keys are requied."
			}, status.HTTP_400_BAD_REQUEST

		x = json_data["x"]
		y = json_data["y"]

		# Check user exists. If so, return conflict
		user = DBUser.query.filter_by(x = x, y = y).first()
		if user:
			return {
				"message": "Conflict. User (%s, %s) exists" % (x, y),
			}, status.HTTP_409_CONFLICT
		# Add user into DB
		user = DBUser(x, y)
		db.session.add(user)
		db.session.flush()

		db.session.commit()
		# Get user ID and return url
		return {
			"message": "Created",
			"user_url": "%s/%s" %(request.url, user.id)
		}, status.HTTP_201_CREATED

class User(Resource):
	"""
	Controller for other user CRUD actions
	Example:
		curl http://127.0.0.1:5000/v1/NN/users/id -X GET
		curl http://127.0.0.1:5000/v1/NN/users/id -X POST -d '{"x": 4}'
		curl http://127.0.0.1:5000/v1/NN/users/id -X DELETE
	"""

	def _not_found_error(self, user_id):
		return {
			"message": "User %s not found" % user_id
		}, status.HTTP_404_NOT_FOUND

	def get(self, user_id):
		"""
		Return User object
		"""
		user = DBUser.query.filter_by(id = user_id).first()
		if not user:
			return self._not_found_error(user_id)
		return {
			"message": "OK",
			"x": user.x,
			"y": user.y
		}, status.HTTP_200_OK
		
	def post(self, user_id):
		"""
		Update User object
		"""
		user = DBUser.query.filter_by(id = user_id).first()
		if not user:
			return self._not_found_error(user_id)
		json_data = request.get_json(force = True)
		if "x" not in json_data and "y" not in json_data:
			return {
				"message": "Bad request. x or y keys are requied."
			}, status.HTTP_400_BAD_REQUEST

		x = json_data.get("x")
		y = json_data.get("y")
		if x:
			user.x = x
		if y:
			user.y = y
		db.session.flush()
		db.session.commit()
		return {
			"message": "OK",
			"user_url": "%s/%s" %(request.url, user.id)
		}, status.HTTP_200_OK
		
	def delete(self, user_id):
		"""
		Delete User object
		"""
		query = DBUser.query.filter_by(id = user_id)
		user = query.first()
		if not user:
			return self._not_found_error(user_id)
		query.delete()
		db.session.commit()
		return {
			"message": "OK"
		}, status.HTTP_200_OK

class Knn(Resource):
	"""
	Controller to find K nearest neighbors
	R (raduis) and U (user_id) arguments are mandatory
	Example:
		curl http://127.0.0.1:5000/v1/NN/users/knn?U=10&R=10 -X GET
	"""
	def __init__(self):
		"""
		Initial conditions:
		- user coord
		- radius
		"""
		self.x0 = None
		self.y0 = None
		self.r = None

	def getMinMaxRectDist(self, stats):
		"""
		Returns min and max distances from point to rectangle
		"""
		dists = list()

		# Add rect coner distances
		for x, y in stats.rect:
			dist = sqrt((self.x0 - x) ** 2 + (self.y0 - y) ** 2)
			dists.append(dist)

		# Add rect side distances
		if (stats.minX <= self.x0 <= stats.maxX):
			for y in (stats.minY, stats.maxY):
				dist = abs(y - self.y0)
				dists.append(dist)
		if (stats.minY <= self.y0 <= stats.maxY):
			for x in (stats.minX, stats.maxX):
				dist = abs(x - self.x0)
				dists.append(dist)

		return min(dists), max(dists)

	def getDistkNN(self, stats):
		"""
		Algorythm by comparing all distances with radius
		Used in benchmark test
		"""
		result = 0
		for user in DBUser.query.yield_per(100):
			dist = sqrt((user.x - self.x0) ** 2 + (user.y - self.y0) ** 2)
			if dist <= self.r:
				result += 1
		return result

	def getkNN(self, stats):
		"""
		=== Main algorythm ===
		If rect small enough (side < R/10) or
		amount of users is small enogh check all distances
		and return users count inside the search zone
		Check rect and the search zone
			- If outside return 0
			- If inside return rect user count
			- If has intersections:
			split into two rect and apply the same algorythm
		Intersections logic:
			- Find min and max distance from initial user to rect
			- Compare distance with R
		Split logic:
			- Split longer rect side.
			- Split by neighbors of avarage value
		"""
		result = 0
		if stats.count == 0:
			return 0

		# Check rectangle is outside, inside or has intersections
		min_dist, max_dist = self.getMinMaxRectDist(stats)
		if (min_dist > self.r) and (max_dist > self.r):
			# Rect outside
			return 0
		elif (min_dist <= self.r) and (max_dist <= self.r):
			# Rect inside
			return stats.count

		# Intersection between rectangle and search area
		if (stats.maxX - stats.minX) < self.r / 10 \
						or (stats.maxY - stats.minY) < self.r \
						or stats.count < MIN_USERS:
			# In case of small amount of users, calculate distances manually
			# In case of small rect side, calculate distances manually
			users = DBUser.query.filter(\
				DBUser.x >= stats.minX, \
				DBUser.x <= stats.maxX, \
				DBUser.y >= stats.minY, \
				DBUser.y <= stats.maxY).all()
			for user in users:
				dist = sqrt((self.x0 - user.x) ** 2 + (self.y0 - user.y) ** 2)
				if dist <= self.r:
					result += 1
		else:
			# Split rect into two in longer side
			if abs(stats.maxX - stats.minX) >= abs(stats.maxY - stats.minY):
				# find nearest left and right of avgX
				leftX = DBUser.query.filter(DBUser.x <= stats.avgX).order_by(db.desc(DBUser.x)).first().x
				rightX = DBUser.query.filter(DBUser.x > stats.avgX).order_by(DBUser.x).first().x
				stats1 = DBUserStats(stats.minX, stats.minY, leftX, stats.maxY)
				stats2 = DBUserStats(rightX, stats.minY, stats.maxX, stats.maxY)
			else:
				# find nearest left and right of avgY
				downY = DBUser.query.filter(DBUser.y <= stats.avgY).order_by(db.desc(DBUser.y)).first().y
				upY = DBUser.query.filter(DBUser.y > stats.avgY).order_by(DBUser.y).first().y
				stats1 = DBUserStats(stats.minX, stats.minY, stats.maxX, downY)
				stats2 = DBUserStats(stats.minX, upY, stats.maxX, stats.maxY)
			result += self.getkNN(stats1)
			result += self.getkNN(stats2)

		return result

	def get(self):
		"""
		Return result of kNN algorythm
		R and U arguments are mandatory
		"""
		r = int(request.args.get('R', 0))
		user_id = int(request.args.get('U', 0))
		dist_angorythm = request.args.get('dist', None)
		if not r:
			return {
				"message": "Bad request. R argument is required."
			}, status.HTTP_400_BAD_REQUEST
		if not user_id:
			return {
				"message": "Bad request. U argument is required."
			}, status.HTTP_400_BAD_REQUEST

		u = DBUser.query.filter_by(id = user_id).first()
		if not u:
			return {
				"message": "User %s not found" % user_id
			}, status.HTTP_404_NOT_FOUND

		dstats = DBUserStats()
		nnstats = DBUserStats(u.x - r, u.y - r, u.x + r, u.y + r)
		init_rect = (
			max(dstats.minX, nnstats.minX),
			max(dstats.minY, nnstats.minY),
			min(dstats.maxX, nnstats.maxX),
			min(dstats.maxY, nnstats.maxY),
		)
		init_stats = DBUserStats(*init_rect)

		self.x0, self.y0 = (u.x, u.y)
		self.r = r

		if dist_angorythm == "Y":
			result = self.getDistkNN(init_stats) - 1
		else:
			result = self.getkNN(init_stats) - 1

		return {
			"message": "OK",
			"result": result,
		}, status.HTTP_200_OK

api.add_resource(UserList, "%s/users" % BASEURL)
api.add_resource(Info, "%s/users/info" % BASEURL)
api.add_resource(User, "%s/users/<int:user_id>" % BASEURL)
api.add_resource(Knn, "%s/users/knn" % BASEURL)

if __name__ == '__main__':
	app.run(debug = True)
